src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadFix,
+                                ct => FixCodeOneParameter(context.Document, context.Span, ct)),
+                            diagnostic);
+                }
+                else
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadFix,
+                                ct => FixCodeTwoParameters(context.Document, context.Span, ct)),
+                            diagnostic);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix =  semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+            // generate text.Length (or whatever "text" is instead in the actual code)
+            var lengthNode = generator
+                .MemberAccessExpression(
+                    fixComponents.StringArgument.Syntax, 
+                    generator.IdentifierName(nameof(string.Length)));
+
+            var startIndexNode = fixComponents.OriginalInnerArguments[0].Value.Syntax;
+
+            // generate "text".Length - 2 (where 2 is the start index given in the original code
+            var lengthArgument = generator
+                .SubtractExpression(
+                    lengthNode,
+                    startIndexNode);
+
+            // generate sb.Append(text, 2, text.Length - 2)
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod, 
+                fixComponents.StringArgument.Syntax,
+                startIndexNode, 
+                lengthArgument);
```

Comment:
do we have tests where the start index is a more complex expression?

Reply:
note: there is a concern here in htat if the start expressoin is side-affecting, it will now execute twice.  probably not significant enough to are about.